include/kernel.h

```diff
@@ -4340,6 +4340,30 @@ extern void k_mem_domain_remove_thread(k_tid_t thread);
  */
 __syscall void k_str_out(char *c, size_t n);
 
+/**
+ * @brief Start a numbered CPU on a MP-capable system
+
+ * This starts and initializes a specific CPU.  The main thread on
+ * startup is running on CPU zero, other processors are numbered
+ * sequentially.  On return from this function, the CPU is known to
+ * have begun operating and will enter the provided function.  Its
+ * interrupts will be initialied but disabled such that irq_unlock()
+ * with the provided key will work to enable them.
+ *
+ * Normally, in SMP mode this function will be called by the kernel
+ * initialization and should not be used as a user API.  But it is
+ * defined here for special-purpose apps which want Zephyr running on
+ * one core and to use others for design-specific processing.
+ *
+ * @param cpu_num Integer number of the CPU
+ * @param stack Stack memory for the CPU
+ * @param sz Stack buffer size, in bytes
+ * @param fn Function to begin running on the CPU.  First argument is an irq_unlock() key.
+ * @param arg Untyped argument to be passed to "fn"
+ */
+extern void _arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz,
```

Comment:
if there are use-cases for this to be invoked from application code, even if only in edge cases, then I think we shouldn't make this a private API, as users would be upset if it changed without warning,

also this worries me a little that this function starts the CPU at a particular memory address, with a stack, without an associated thread object. we are basically running the CPU without being in the context of a thread, something that doesn't happen in Zephyr except in the very early boot stages. if you tried to use some kernel objects in this weird context-free state I don't know what happens, or what would happen if this context-free state was swapped out.

is there some way we can integrate this with k_thread_create() instead, at least for the user-facing scenario? for example you could perhaps reserve some bits in the k_thread_create() options field to specify the CPU number or something.

Reply:
This is the lower level primitive that is used by the SMP layer to implement threading.  Something like this has to be there, if you think about it.  Whether or not it's worth documenting as a public API or not is an open question.  I chose to document it, since asymmetric MP is a real thing and some apps might want to have it in a circumstance where they're using Zephyr to manage CPU0.

Reply:
OK let's leave this as private then, people who do decide to use it in an app will be assumed to know exactly what they are doing.